cumulus/client/consensus/aura/src/collators/mod.rs

```diff
@@ -270,6 +281,177 @@ where
+	) -> Self {
+		Self {
+			inner,
+			client,
+			create_inherent_data_providers,
+			check_for_equivocation,
+			compatibility_mode,
+			_phantom: Default::default(),
+		}
+	}
+}
+
+#[async_trait]
+impl<Block, BI, Client, CIDP, P> BlockImport<Block>
+	for ValidatingBlockImport<BI, Client, CIDP, NumberFor<Block>, P>
+where
+	Block: BlockT,
+	BI: BlockImport<Block> + Sync,
+	BI::Error: Into<sp_consensus::Error>,
+	Client: ProvideRuntimeApi<Block> + sc_client_api::backend::AuxStore + Send + Sync,
+	Client::Api: Core<Block> + BlockBuilder<Block> + AuraApi<Block, <P as Pair>::Public>,
+	P: Pair + Sync,
+	P::Public: Codec + std::fmt::Debug,
+	P::Signature: Codec,
+	CIDP: sp_inherents::CreateInherentDataProviders<Block, ()> + Send,
+	CIDP::InherentDataProviders: InherentDataProviderExt + Send + Sync,
+{
+	type Error = sp_consensus::Error;
+
+	async fn check_block(
+		&self,
+		block: BlockCheckParams<Block>,
+	) -> Result<ImportResult, Self::Error> {
+		self.inner.check_block(block).await.map_err(Into::into)
+	}
+
+	async fn import_block(
+		&self,
+		mut block: BlockImportParams<Block>,
+	) -> Result<ImportResult, Self::Error> {
+		validate_block_import::<_, _, P, _>(
+			&mut block,
+			self.client.as_ref(),
+			&self.create_inherent_data_providers,
+			self.check_for_equivocation,
+			&self.compatibility_mode,
+		)
+		.await?;
+		self.inner.import_block(block).await.map_err(Into::into)
+	}
+}
+
+impl<BI, Client, CIDP, N, P> ParachainBlockImportMarker
+	for ValidatingBlockImport<BI, Client, CIDP, N, P>
+{
+}
+
+async fn validate_block_import<Block, Client, P, CIDP>(
+	params: &mut sc_consensus::BlockImportParams<Block>,
+	client: &Client,
+	create_inherent_data_providers: &CIDP,
+	check_for_equivocation: bool,
+	compatibility_mode: &CompatibilityMode<NumberFor<Block>>,
+) -> Result<(), sp_consensus::Error>
+where
+	Block: BlockT,
+	Client: ProvideRuntimeApi<Block> + sc_client_api::backend::AuxStore + Send + Sync,
+	Client::Api: Core<Block> + BlockBuilder<Block> + AuraApi<Block, <P as Pair>::Public>,
+	P: Pair + Sync,
+	P::Public: Codec + std::fmt::Debug,
+	P::Signature: Codec,
+	CIDP: sp_inherents::CreateInherentDataProviders<Block, ()> + Send,
+	CIDP::InherentDataProviders: InherentDataProviderExt + Send + Sync,
+{
+	// Check inherents.
+	let slot = find_pre_digest::<Block, P::Signature>(&params.header)
+		.map_err(|e| sp_consensus::Error::Other(Box::new(e)))?;
+	let parent_hash = *params.header.parent_hash();
+	// if the body is passed through, we need to use the runtime
+	// to check that the internally-set timestamp in the inherents
+	// actually matches the slot set in the seal.
+	if let Some(inner_body) = params.body.take() {
+		let new_block = Block::new(params.header.clone(), inner_body);
+		if client
+			.runtime_api()
+			.has_api_with::<dyn BlockBuilder<Block>, _>(parent_hash, |v| v >= 2)
+			.map_err(|e| sp_consensus::Error::Other(Box::new(e)))?
+		{
+			let create_inherent_data_providers = create_inherent_data_providers
```

Comment:
I think we can move this out of the `if` blocks, to be used in the check for equivocation part too.